src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.SymbolStartAnalyzer.BlockAnalyzer.cs

```diff
@@ -0,0 +1,492 @@
+                    {
+                        return;
+                    }
+
+                    var expressionStatement = (IExpressionStatementOperation)context.Operation;
+                    var value = expressionStatement.Operation;
+
+                    // Bail out cases for report unused expression value:
+
+                    //  1. Null type and void returning method invocations: no value being dropped here.
+                    if (value.Type == null ||
+                        value.Type.SpecialType == SpecialType.System_Void)
+                    {
+                        return;
+                    }
+
+                    //  2. Bail out for syntax error (constant expressions) and semantic error (invalid operation) cases.
+                    if (value.ConstantValue.HasValue ||
+                        value is IInvalidOperation)
+                    {
+                        return;
+                    }
+
+                    //  3. Assignments, increment/decrement operations: value is actually being assigned.
+                    if (value is IAssignmentOperation ||
+                        value is IIncrementOrDecrementOperation)
+                    {
+                        return;
+                    }
+
+                    //  4. Bool returning method invocations: these are extremely noisy as large number of bool 
+                    //     returning methods return the status of the operation.
+                    //     In future, if we feel noise level is fine, we can consider removing this check.
+                    if (value.Type.SpecialType == SpecialType.System_Boolean)
+                    {
+                        return;
+                    }
+
+                    //  5. If expression statement and its underlying expression have differing first tokens that likely indicates
+                    //     an explicit discard. For example, VB call statement is used to explicitly ignore the value returned by
+                    //     an invocation by prefixing the invocation with keyword "Call".
+                    if (value.Syntax.GetFirstToken() != expressionStatement.Syntax.GetFirstToken())
+                    {
+                        return;
+                    }
+
+                    //  6. Special cases where return value is not required to be checked:
+                    //     Methods belonging to System.Threading.Interlocked and System.Collections.Immutable.ImmutableInterlocked
+                    //     return the original value, which is not required to be checked.
+                    if (value is IInvocationOperation invocation &&
+                        (invocation.TargetMethod.ContainingType.OriginalDefinition == _symbolStartAnalyzer._interlockedTypeOpt ||
+                         invocation.TargetMethod.ContainingType.OriginalDefinition == _symbolStartAnalyzer._immutableInterlockedTypeOpt))
+                    {
+                        return;
+                    }
```

Comment:
this worries me a lot.  We're basically signing up to have total knowledge of hte set of methods where it's reasonable to assume we know the return value isn't needed.

Reply:
Actually, given this is now a hidden diagnostic/refactoring with no user visible impact, we really don't need to worry about these special cases. I am going to remove the special bail out cases here, which were added for original design.

---
In reply to: [234802838](https://github.com/dotnet/roslyn/pull/30777#discussion_r234802838) [](ancestors = 234802838)